File: example/java/frugal-lib/com/workiva/frugal/NatsThriftTransport.java
package com.workiva.frugal;

import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;
import org.nats.Connection;
import org.nats.MsgHandler;

import java.io.IOException;
import java.io.PipedInputStream;
import java.io.PipedOutputStream;
import java.nio.ByteBuffer;
import java.util.Arrays;

/**
 * NatsThriftTransport is an extension of thrift.TTransport exclusively used for
 * pub/sub via NATS.
 */
public class NatsThriftTransport extends TTransport {

    // NATS limits messages to 1MB.
    private static final int NATS_MAX_MESSAGE_SIZE = 1024 * 1024;
    private static final int UNSUBSCRIBED = -1;

    private Connection conn;
    private PipedOutputStream writer;
    private PipedInputStream reader;
    private ByteBuffer writeBuffer;
    private int sub;
    private String subject;

    public NatsThriftTransport(Connection conn) {
        this.conn = conn;
        writeBuffer = ByteBuffer.allocate(NATS_MAX_MESSAGE_SIZE);
        sub = UNSUBSCRIBED;
    }

    @Override
    public boolean isOpen() {
        return sub != UNSUBSCRIBED;
    }

    @Override
    public void open() throws TTransportException {
        try {
            writer = new PipedOutputStream();
            reader = new PipedInputStream(writer);
        } catch (IOException e) {
            throw new TTransportException(TTransportException.UNKNOWN, e);
        }

        try {
            sub = conn.subscribe(subject, new MsgHandler() {
                @Override
                public void execute(byte[] msg, String reply, String subject) {
                    try {
                        writer.write(msg);
                    } catch (IOException e) {
                        // TODO: What do we do here?
                        e.printStackTrace();
                    }
                }
            });
        } catch (IOException e) {
            throw new TTransportException(TTransportException.UNKNOWN, e);
        }
    }

    @Override
    public void close() {
        if (!isOpen()) {
            return;
        }
        try {
            conn.unsubscribe(sub);
        } catch (IOException e) {
            // TODO: What do we do here?
            e.printStackTrace();
        }
        sub = UNSUBSCRIBED;
        try {
            writer.close();
        } catch (IOException e) {
            // TODO: What do we do here?
            e.printStackTrace();
        }
    }

    @Override
    public int read(byte[] bytes, int off, int len) throws TTransportException {
        try {
            return reader.read(bytes, off, len);
        } catch (IOException e) {
            throw new TTransportException(TTransportException.UNKNOWN, e);
        }
    }

    @Override
    public void write(byte[] bytes, int off, int len) throws TTransportException {
        if (bytes.length + writeBuffer.position() > writeBuffer.capacity()) {
            writeBuffer.clear();
            throw new TTransportException(TTransportException.UNKNOWN,
                    "Message is too large");
        }
        writeBuffer.put(bytes, off, len);
    }

    @Override
    public void flush() throws TTransportException {
        byte[] data = new byte[writeBuffer.position()];
        writeBuffer.flip();
        writeBuffer.get(data);
        if (data.length == 0) {
            return;
        }
        try {
            conn.publish(subject, null, data, null);
        } catch (IOException e) {
            throw new TTransportException(TTransportException.UNKNOWN, e);
        } finally {
            writeBuffer.clear();
        }
    }

    public void setSubject(String subject) {
        this.subject = subject;
    }

}

File: Godeps/_workspace/src/github.com/Workiva/frugal-go/nats_thrift_transport.go
package frugal

import (
	"bufio"
	"bytes"
	"errors"
	"io"

	"git.apache.org/thrift.git/lib/go/thrift"
	"github.com/nats-io/nats"
)

var (
	ErrTooLarge     = errors.New("Message is too large")
	ErrEmptyMessage = errors.New("Message is empty")
)

// NATS limits messages to 1MB.
const natsMaxMessageSize = 1024 * 1024

// natsThriftTransport is an implementation of thrift.TTransport exclusively
// used for pub/sub via NATS.
type natsThriftTransport struct {
	conn        *nats.Conn
	subject     string
	reader      *bufio.Reader
	writer      *io.PipeWriter
	writeBuffer *bytes.Buffer
	sub         *nats.Subscription
}

func newNATSThriftTransport(conn *nats.Conn) *natsThriftTransport {
	buf := make([]byte, 0, natsMaxMessageSize)
	return &natsThriftTransport{
		conn:        conn,
		writeBuffer: bytes.NewBuffer(buf),
	}
}

func (n *natsThriftTransport) Open() error {
	reader, writer := io.Pipe()
	n.reader = bufio.NewReader(reader)
	n.writer = writer
	sub, err := n.conn.Subscribe(n.subject, func(msg *nats.Msg) {
		n.writer.Write(msg.Data)
	})
	if err != nil {
		return thrift.NewTTransportExceptionFromError(err)
	}
	n.sub = sub
	return nil
}

func (n *natsThriftTransport) IsOpen() bool {
	return n.sub != nil
}

func (n *natsThriftTransport) Close() error {
	if !n.IsOpen() {
		return nil
	}
	if err := n.sub.Unsubscribe(); err != nil {
		return err
	}
	n.sub = nil
	return thrift.NewTTransportExceptionFromError(n.writer.Close())
}

func (n *natsThriftTransport) Read(p []byte) (int, error) {
	num, err := n.reader.Read(p)
	return num, thrift.NewTTransportExceptionFromError(err)
}

func (n *natsThriftTransport) Write(p []byte) (int, error) {
	if len(p)+n.writeBuffer.Len() > natsMaxMessageSize {
		n.writeBuffer.Reset() // Clear any existing bytes.
		return 0, ErrTooLarge
	}
	num, err := n.writeBuffer.Write(p)
	return num, thrift.NewTTransportExceptionFromError(err)
}

func (n *natsThriftTransport) Flush() error {
	data := n.writeBuffer.Bytes()
	if len(data) == 0 {
		return ErrEmptyMessage
	}
	err := n.conn.Publish(n.subject, data)
	n.writeBuffer.Reset()
	return thrift.NewTTransportExceptionFromError(err)
}

func (n *natsThriftTransport) RemainingBytes() uint64 {
	return ^uint64(0) // We don't know unless framed is used. See thrift-nats impl.
}

func (n *natsThriftTransport) SetSubject(subject string) {
	n.subject = subject
}
